etl-destinations/src/bigquery/core.rs

```diff
@@ -183,6 +200,89 @@ where
         Ok((table_id, table_descriptor))
     }
 
+    /// Adds a table ID to the created tables cache with random eviction if full.
+    ///
+    /// If the cache is at maximum size, removes a random table before adding the new one.
+    fn add_to_created_tables_cache(inner: &mut Inner<impl SchemaStore>, table_id: BigQueryTableId) {
+        // If table is already in cache, nothing to do
+        if inner.created_tables.contains(&table_id) {
+            return;
+        }
+
+        // If cache is full, evict a random item
+        if inner.created_tables.len() >= MAX_CREATED_TABLES_CACHE_SIZE {
+            // Get a random table from the set to remove
+            if let Some(random_table_id) = inner.created_tables.iter().next().cloned() {
+                inner.created_tables.remove(&random_table_id);
+                info!("evicted table {random_table_id} from creation cache (cache full)");
+            }
+        }
+
+        // Add the new table to cache
+        inner.created_tables.insert(table_id);
+    }
+
+    /// Removes a table ID from the created tables cache.
+    ///
+    /// Used when a table is found to not exist during streaming operations.
+    fn remove_from_created_tables_cache(
+        inner: &mut Inner<impl SchemaStore>,
+        table_id: &BigQueryTableId,
+    ) {
+        inner.created_tables.remove(table_id);
+    }
+
+    /// Handles streaming rows with fallback table creation on missing table errors.
+    ///
+    /// Attempts to stream rows to BigQuery. If streaming fails with a table not found error,
+    /// removes the table from the cache and retries with table creation.
+    async fn stream_rows_with_fallback(
+        inner: &mut Inner<S>,
+        dataset_id: &BigQueryDatasetId,
+        table_id: &BigQueryTableId,
+        table_descriptor: &TableDescriptor,
+        table_rows: Vec<TableRow>,
+        orig_table_id: &TableId,
+        use_cdc_sequence_column: bool,
+    ) -> EtlResult<()> {
+        // First attempt - optimistically assume the table exists
+        let result = inner
+            .client
+            .stream_rows(dataset_id, table_id, table_descriptor, table_rows.clone())
+            .await;
+
+        match result {
```

Comment:
We try only one time, if it fails, we propagate the error. We can't afford to retry multiple times on such a hot path.